File: team.py
from tfmatch import MatchResult
from functools import total_ordering


@total_ordering
class Team:
    def __init__(self, name: str, skill: float, bye=False):
        """
        Initializes a team
        :param name:
        :param skill:
        """
        self.name = name
        self.skill = skill
        self.bye = bye

    def __lt__(self, other):
        return self.skill < other.skill

    def __eq__(self, other):
        return self.skill == other.skill


class ByeTeam(Team):
    def __init__(self):
        super().__init__("Bye Week", -99999999, True)


@total_ordering
class TeamContext:
    def __init__(self, team: Team):
        """
        Gives a team its context in a division,
        :param team: the Team object
        """
        self.team = team
        self.wins = 0
        self.losses = 0
        self.rounds_won = 0
        self.rounds_lost = 0
        self.match_points = 0
        self.teams_faced = []

    def __add__(self, other: MatchResult):
        self.wins += (1 if other.won else 0)
        self.losses += (0 if other.won else 1)
        self.rounds_won += other.rounds_won
        self.rounds_lost += other.rounds_lost
        self.match_points += other.match_points

    # Returns, tab separated: name, skill, wins, losses, rounds won, rounds lost, match points, inq match points
    def __str__(self):
        return self.team.name + "," + str(round(self.team.skill, 3)) + "," + str(self.wins) + "," + str(self.losses) \
               + "," + str(self.rounds_won) + "," + str(self.rounds_lost) + "," + str(self.match_points)

    def __repr__(self):
        return self.team.name

    def __lt__(self, other):
        if self.match_points < other.match_points:
            return True
        elif self.match_points > other.match_points:
            return False
        # Tiebreak by Median Buchholz
        median_buchholz = 0
        max_value = 0
        min_value = 999999 #Really Big Number
        for team in self.teams_faced:
            median_buchholz += team.match_points
            max_value = max(max_value, team.match_points)
            min_value = min(min_value, team.match_points)
        median_buchholz -= max_value
        median_buchholz -= min_value
        return max(median_buchholz, 0)

File: src/main/java/com/vibeisveryo/RGLHighlanderMatchPointSimulation/Team.java
package com.vibeisveryo.RGLHighlanderMatchPointSimulation;

import java.util.ArrayList;
import java.util.HashSet;


public class Team implements Comparable<Team> {
    private String name;
    private double skill;
    private boolean bye;


    private int wins;
    private int losses;
    private int roundsWon;
    private int roundsLost;
    private int matchPoints;
    private HashSet<Team> teamsFaced;

    public Team(String name, double skill) {
        this.name = name;
        this.skill = skill;
        this.bye = false;
        this.wins = 0;
        this.losses = 0;
        this.roundsWon = 0;
        this.roundsLost = 0;
        this.matchPoints = 0;
        this.teamsFaced = new HashSet<>();
    }

    public String getName() {
        return name;
    }

    public double getSkill() {
        return skill;
    }

    public boolean isBye() {
        return bye;
    }

    public void addMatch(MatchResult matchResult) {
        this.wins += (matchResult.won() ? 1 : 0);
        this.losses += (matchResult.won() ? 0 : 1);
        this.roundsWon += matchResult.roundsWon();
        this.roundsLost += matchResult.roundsLost();
        this.matchPoints += matchResult.matchPoints();
    }

    public String toString() {
        return this.getName() + "," + String.format("%.3f", this.getSkill()) + "," + this.wins + "," + this.losses
                + "," + this.roundsWon + "," + this.roundsLost + "," + this.matchPoints;
    }

    public int compareTo(Team other) {
        if (this.matchPoints < other.matchPoints) {
            return -1;
        } else if (this.matchPoints > other.matchPoints) {
            return 1;
        } else {
            // Tiebreak by Median Buchholz
            int thisMB = 0;
            int otherMB = 0;
            {
                int max = 0;
                int min = Integer.MAX_VALUE; // Really Big Number
                for (Team team: this.teamsFaced) {
                    thisMB += team.matchPoints;
                    max = Math.max(max, team.matchPoints);
                    min = Math.min(min, team.matchPoints);
                }
                thisMB = thisMB - max - min;
            }
            {
                int max = 0;
                int min = Integer.MAX_VALUE; // Really Big Number
                for (Team team: other.teamsFaced) {
                    otherMB += team.matchPoints;
                    max = Math.max(max, team.matchPoints);
                    min = Math.min(min, team.matchPoints);
                }
                otherMB = otherMB - max - min;
            }
            return thisMB - otherMB;
        }
    }

    public Team getTeam() {
        return this;
    }

    public int getWins() {
        return wins;
    }

    public int getLosses() {
        return losses;
    }

    public int getRoundsWon() {
        return roundsWon;
    }

    public int getRoundsLost() {
        return roundsLost;
    }

    public int getMatchPoints() {
        return matchPoints;
    }

    public HashSet<Team> getTeamsFaced() {
        return teamsFaced;
    }

    public ArrayList<String> getTeamsFacedNames() {
        ArrayList<String> list = new ArrayList<>();
        for (Team team: this.teamsFaced) {
            list.add(team.getName());
        }
        return list;
    }

    public void addTeamFaced(Team team) {
        teamsFaced.add(team);
    }
}

class ByeTeam extends Team {
    boolean bye;
    public ByeTeam() {
        super("Bye Week", -99999999);
        this.bye = true;
    }
}

